core/src/main/java/io/questdb/cairo/O3PurgeDiscoveryJob.java

```diff
@@ -28,181 +28,266 @@
-import io.questdb.tasks.O3PurgeTask;
-import org.jetbrains.annotations.Nullable;
 
-public class O3PurgeDiscoveryJob extends AbstractQueueConsumerJob<O3PurgeDiscoveryTask> {
+import java.io.Closeable;
+import java.io.IOException;
+
+public class O3PurgeDiscoveryJob extends AbstractQueueConsumerJob<O3PurgeDiscoveryTask> implements Closeable {
 
     private final static Log LOG = LogFactory.getLog(O3PurgeDiscoveryJob.class);
     private final CairoConfiguration configuration;
     private final MutableCharSink[] sink;
     private final StringSink[] fileNameSinks;
-    private final LongList[] txnList;
-    private final RingQueue<O3PurgeTask> purgeQueue;
-    private final Sequence purgePubSeq;
+    private final ObjList<DirectLongList> partitionList;
+    private final ObjList<TxnScoreboard> txnScoreboards;
+    private final ObjList<TxReader> txnReaders;
 
     public O3PurgeDiscoveryJob(MessageBus messageBus, int workerCount) {
         super(messageBus.getO3PurgeDiscoveryQueue(), messageBus.getO3PurgeDiscoverySubSeq());
         this.configuration = messageBus.getConfiguration();
-        this.purgeQueue = messageBus.getO3PurgeQueue();
-        this.purgePubSeq = messageBus.getO3PurgePubSeq();
         this.sink = new MutableCharSink[workerCount];
         this.fileNameSinks = new StringSink[workerCount];
-        this.txnList = new LongList[workerCount];
+        this.partitionList = new ObjList<>(workerCount);
+        this.txnScoreboards = new ObjList<>(workerCount);
+        this.txnReaders = new ObjList<>(workerCount);
+
         for (int i = 0; i < workerCount; i++) {
             sink[i] = new StringSink();
             fileNameSinks[i] = new StringSink();
-            txnList[i] = new LongList();
+            partitionList.add(new DirectLongList(configuration.getPartitionPurgeListCapacity() * 2, MemoryTag.NATIVE_LONG_LIST));
+            txnScoreboards.add(new TxnScoreboard(configuration.getFilesFacade(), configuration.getTxnScoreboardEntryCount()));
+            txnReaders.add(new TxReader(configuration.getFilesFacade()));
         }
     }
 
-    public static boolean discoverPartitions(
+    @Override
+    public void close() throws IOException {
+        Misc.freeObjList(partitionList);
+        Misc.freeObjList(txnReaders);
+        Misc.freeObjList(txnScoreboards);
+    }
+
+    private static void processPartition(
+            FilesFacade ff,
+            Path path,
+            int tableRootLen,
+            TxReader txReader,
+            TxnScoreboard txnScoreboard,
+            long partitionTimestamp,
+            int partitionBy,
+            DirectLongList partitionList,
+            int lo,
+            int hi
+    ) {
+        long lastCommittedPartitionName = txReader.getPartitionNameTxnByPartitionTimestamp(partitionTimestamp);
+        if (lastCommittedPartitionName > -1) {
+            assert hi <= partitionList.size();
+            for (int i = lo + 2, n = hi - 1; i < n; i += 2) {
```

Comment:
This needs further explanation:

- why we need formulas
- what are the magic nimbers

Reply:
updated in fixed in https://github.com/questdb/questdb/pull/1778/commits/c07fd7aa136747a7679d3865ffbb164535d0ee3d